docs/2020-06-17-Checking-Implication.md

```diff
@@ -0,0 +1,161 @@
+# Checking Implication
+
+Checking implication is an important step for proving one-path and all-path reachability claims.
+This document will describe logically how this step will be performed.
+
+## Background
+
+The reader should be familiar with the algorithm for proving one-path and all-path reachability claims:
+
+- [One-Path](2018-11-08-One-Path-Reachability-Proofs.md)
+- [All-Path](2019-03-28-All-Path-Reachability-Proofs.md)
+
+## Problem
+
+Consider a reachability claim,
+
+```
+φ(X) → ∘ ∃ Y. ψ(X, Y)
+```
+
+where `∘` denotes the appropriate one-path or all-path modality
+and `X` and `Y` denote _distinct_ families of variables.
+We will assume that `φ(X)` and `ψ(X, Y)` are function-like patterns (matching at most one value)
+because this is true in typical use.
+
+The reachability proof algorithm will unroll the modality, giving
+
+```
+φ(X) → ∃ Y. ψ(X, Y) ∨ • ∘ ∃ Y. ψ(X, Y)
+```
+
+If the direct implication holds,
+
+```
+φ(X) → ∃ Y. ψ(X, Y)
+```
+
+then the claim is proven.
+Otherwise, the algorithm continues with
+
+```
+φ(X) → • ∘ ∃ Y. ψ(X, Y)
```

Comment:
That's not true anymore. At one point in the past, the backend continued with 

```
φ(X) ∧ ¬ ∃ Y. ⌈φ(X) ∧ ψ(X, Y)⌉ → • ∘ ∃ Y. ψ(X, Y).
```

I think that's formally equivalent to what you wrote. The backend now reports any configuration as "stuck" unless the implication condition `¬ ∃ Y. ⌈φ(X) ∧ ψ(X, Y)⌉ ` is exactly `\top` (in which case we keep rewriting) or `\bottom` (in which case the proof is done). I agree that this is technically incomplete, but so far it has served fine. Do you think it is important that we handle this now? If so, I should add it to the document, because there are again important ways to optimize the implementation.

Reply:
Probably not important at the moment; however, we should have a note about it somewhere in case it ever becomes relevant.